service/history/replication/metrics_emitter.go

```diff
@@ -0,0 +1,178 @@
+	shardData MetricsEmitterShardData,
+	reader taskReader,
+	metricsClient metrics.Client,
+) *metricsEmitterImpl {
+	shardID := shardData.GetShardID()
+	currentCluster := shardData.GetClusterMetadata().GetCurrentClusterName()
+
+	remoteClusters := shardData.GetClusterMetadata().GetRemoteClusterInfo()
+	remoteClusterNames := make([]string, len(remoteClusters))
+	for remoteCluster := range remoteClusters {
+		remoteClusterNames = append(remoteClusterNames, remoteCluster)
+	}
+
+	scope := metricsClient.Scope(
+		metrics.ReplicationMetricEmitterScope,
+		metrics.ActiveClusterTag(currentCluster),
+		metrics.InstanceTag(strconv.Itoa(shardID)),
+	)
+	logger := shardData.GetLogger().WithTags(
+		tag.ClusterName(currentCluster),
+		tag.ShardID(shardID))
+
+	return &metricsEmitterImpl{
+		currentCluster:     currentCluster,
+		remoteClusterNames: remoteClusterNames,
+		status:             common.DaemonStatusInitialized,
+		shardData:          shardData,
+		reader:             reader,
+		scope:              scope,
+		logger:             logger,
+		done:               make(chan struct{}),
+	}
+}
+
+func (m *metricsEmitterImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&m.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	go m.emitMetricsLoop()
+	m.logger.Info("ReplicationMetricsEmitter started.")
+}
+
+func (m *metricsEmitterImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&m.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	m.logger.Info("ReplicationMetricsEmitter shutting down.")
+	close(m.done)
+}
+
+func (m *metricsEmitterImpl) emitMetricsLoop() {
+	ticker := time.NewTicker(metricsEmissionInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.done:
+			return
+		case <-ticker.C:
+			m.emitMetrics()
+		}
+	}
+}
+
+func (m *metricsEmitterImpl) emitMetrics() {
+	for _, remoteClusterName := range m.remoteClusterNames {
+		logger := m.logger.WithTags(tag.RemoteCluster(remoteClusterName))
+		scope := m.scope.Tagged(metrics.TargetClusterTag(remoteClusterName))
+
+		replicationLatency, err := m.determineReplicationLatency(remoteClusterName)
+		if err != nil {
+			return
+		}
+
+		scope.UpdateGauge(metrics.ReplicationLatency, float64(replicationLatency.Nanoseconds()))
+		logger.Debug(fmt.Sprintf("ReplicationLatency metric emitted: %v", float64(replicationLatency.Nanoseconds())))
+	}
+}
+
+func (m *metricsEmitterImpl) determineReplicationLatency(remoteClusterName string) (time.Duration, error) {
+	logger := m.logger.WithTags(tag.RemoteCluster(remoteClusterName))
+	lastReadTaskID := m.shardData.GetClusterReplicationLevel(remoteClusterName)
+	maxReadLevel := m.shardData.GetTransferMaxReadLevel()
+
+	// TODO I should probably use task_store instead to get from cache, even though I don't need it hydrated
+	// TODO Do I need to use a different context? Or add a deadline / cancellation?
+	tasks, _, err := m.reader.Read(ctx.Background(), lastReadTaskID, maxReadLevel)
+	if err != nil {
```

Comment:
Correct, you do not need this hydrated. I think we can leave it as is for now.
`context.Background()` should be fine. I do not see much benefit on restricting deadline for such a background processing task.